pages/blog/guide-to-nova-gc.md

```diff
@@ -0,0 +1,678 @@
+`GcScope<'a, '_>`, ie. when a bound value is returned from a function that can
+trigger garbage collection. Due to certain details of Rust's borrow checker
+rules related to internal mutation, the returned value will keep the `GcScope`
+from being reused. We'll see how to deal with this situation later.
+
+## The Practice
+
+Now that we have seen the `GcScope`, `NoGcScope`, and bound values, we are ready
+to start using them in practice. First, let's do a quick review of these three
+principal actors:
+
+1. `GcScope<'a, '_>`: A zero-sized type that is passed to functions that can
+   trigger garbage collection. Only one such type is ever active at a time, and
+   no `NoGcScope`s can be active at the same time.
+2. `NoGcScope<'a, '_>`: A zero-sized type that is passed to functions that
+   cannot trigger garbage collection. Any number of such types can be active at
+   the same time, but no `GcScope` can be active at the same time.
+3. Bound values: Handles to garbage collected values that have been bound to a
+   `NoGcScope` or `GcScope`, implicitly or explicitly.
+
+With this, we're ready to start our praxis.
+
+### Returning bound values from `NoGcScope` functions
+
+The simplest function that deals with bound values we can imagine in the engine
+is one that cannot trigger garbage collection, and returns a bound value. An
+example of this would be `Value::from_string` which takes a Rust
+`std::string::String` by value, moves it to the `Agent` heap, and returns a
+`Value` handle to it.
+
+```rust
+pub fn from_string<'a>(
+    agent: &mut Agent,
+    string: std::string::String,
+    gc: NoGcScope<'a, '_>
+) -> Value<'a>;
+```
+
+There is nothing much to say about these sorts of functions: The only thing of
+note is that it is important for the `'a` lifetime to be defined and used in
+`NoGcScope<'a, '_>`. The following would be _incorrect_:
+
+```rust
+// *Incorrect* usage!
+pub fn from_string_wrong(
+    agent: &mut Agent,
+    string: std::string::String,
+    gc: NoGcScope
+) -> Value; // Wrong! Lifetime not bound to NoGcScope!
+```
+
+With this sort of definition, the returned `Value` would be bound to the implied
+`'a` lifetime of `&'a mut Agent`, not to the `NoGcScope` lifetime. This would
+not be a "bound value". This would also block the entire `Agent` from being used
+while the returned `Value` still exists. Even if the blocking were to be fixed
+(by making the function take `&Agent` instead), it would still mean that the
+handle would be invalidated if the `Agent` is mutated. This isn't correct: We
+want the `Value` to be invalidated when garbage collection may be triggered,
+which can happen entirely unconnected from heap mutation.
+
+### Returning bound values from `GcScope` functions
+
+Functions that can trigger garbage collection, return bound values, and don't
+take any parameters do not actually exist in the engine at all but if they did,
+they would look like this:
```

Comment:
The "do not actually exist in the engine at all" part makes it sound like this is a counterfactual or something. Reword to "there are no such functions currently" or something like that, if that is the intention.

Reply:
Fixed, thanks <3